### PR TITLE
Issue #2296 - implement server resolve function

### PR DESCRIPTION
**Pull request by JohnTimm**
Signed-off-by: John T.E. Timm <johntimm@us.ibm.com>

### CODE REVIEW

File: fhir-model/src/test/java/com/ibm/fhir/model/util/test/ReferencePatternTest.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ */
+
+package com.ibm.fhir.model.util.test;
+
+import static com.ibm.fhir.model.util.FHIRUtil.REFERENCE_PATTERN;
+
+import java.util.regex.Matcher;
+
+public class ReferencePatternTest {
+    public static void main(String[] args) throws Exception {
```
Comment (prb112) [inline fhir-model/src/test/java/com/ibm/fhir/model/util/test/ReferencePatternTest.java:14]: Worthwhile making a Test since it'd be a good confirmation.

**Review by prb112 — APPROVED** [review]
LGTM (Pending the docs)

File: fhir-model/src/test/java/com/ibm/fhir/model/util/test/ReferencePatternTest.java (hunk 1 above)
Comment (JohnTimm) [reply fhir-model/src/test/java/com/ibm/fhir/model/util/test/ReferencePatternTest.java]: Done

**Comment by lmsurpre** [comment]
should we consider resolving absolute references as well?

how about absolute references with a base that matches the configured server base url?

**Comment by JohnTimm** [comment]
> should we consider resolving absolute references as well?

> how about absolute references with a base that matches the configured server base url?

Yes we could consider absolute references but only in the context of the ServerResolveFunction. The base ResolveFunction in the fhir-path module has no way of knowing the server base URL.

> should we try to consolidate any of the reference parsing logic with what we have in ReferenceUtil?

Perhaps, however, ReferenceUtil is in fhir-search which is not a dependency of fhir-path. I’m good keeping them separate for now and leaving reconciliation for future work.

File: fhir-server/src/test/java/com/ibm/fhir/server/test/ServerResolveFunctionTest.java [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+            return persistence;
+        }
+
+        @Override
+        public FHIRPersistence getFHIRPersistenceImplementation(String factoryPropertyName) throws FHIRPersistenceException {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    public static class PersistenceImpl implements FHIRPersistence {
```
Comment (lmsurpre) [inline fhir-server/src/test/java/com/ibm/fhir/server/test/ServerResolveFunctionTest.java:208]: This is like the 4th mock / stubbed out persistence impl we have across our various modules.  I wonder if its worth trying to consolidate.

File: fhir-config/src/main/java/com/ibm/fhir/config/FHIRConfiguration.java [hunk 3]
```diff
@@ -33,6 +33,7 @@
     public static final String PROPERTY_CHECK_REFERENCE_TYPES = "fhirServer/core/checkReferenceTypes";
     public static final String PROPERTY_CONDITIONAL_DELETE_MAX_NUMBER = "fhirServer/core/conditionalDeleteMaxNumber";
     public static final String PROPERTY_SERVER_REGISTRY_RESOURCE_PROVIDER_ENABLED = "fhirServer/core/serverRegistryResourceProviderEnabled";
+    public static final String PROPERTY_SERVER_RESOLVE_FUNCTION_ENABLED = "fhirServer/core/serverResolveFunctionEnabled";
```
Comment (lmsurpre) [inline fhir-config/src/main/java/com/ibm/fhir/config/FHIRConfiguration.java:36]: as paul mentioned, we need to document this on in the user's guide

File: fhir-server/src/test/java/com/ibm/fhir/server/test/ServerResolveFunctionTest.java (hunk 2 above)
Comment (JohnTimm) [reply fhir-server/src/test/java/com/ibm/fhir/server/test/ServerResolveFunctionTest.java:208]: Yes, the one I made is map-based and could be generalized and perhaps reused elsewhere. That said, such consolidation is outside the scope of this PR.

File: fhir-config/src/main/java/com/ibm/fhir/config/FHIRConfiguration.java (hunk 3 above)
Comment (JohnTimm) [reply fhir-config/src/main/java/com/ibm/fhir/config/FHIRConfiguration.java:36]: Users Guide updates are forthcoming.

Comment (JohnTimm) [reply fhir-config/src/main/java/com/ibm/fhir/config/FHIRConfiguration.java:36]: Updated FHIR Server User's Guide with information on the new configuration property.

**Review by lmsurpre — APPROVED** [review]
LGTM

**Pull request merged**